### PR TITLE
feat(certificates) rework vault certificate provider 

### PR DESCRIPTION
**Pull request by schristoff**
Builds on the work done in #4580 - this removes all of the caching, messaging broker and more to allow the Vault Provider to only focus on certificates. 
(accidentally put this in the wrong block so was hidden, oops!) 

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| CI System                  | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [X ] |
| Control Plane              | [ ] |
| Demo                       | [ ] |
| Documentation              | [ ] |
| Egress                     | [ ] |
| Ingress                    | [ ] |
| Install                    | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| Performance                | [ ] |
| SMI Policy                 | [ ] |
| Security                   | [ ] |
| Sidecar Injection          | [ ] |
| Tests                      | [ ] |
| Upgrade                    | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?
no

2. Is this a breaking change?
no

### CODE REVIEW

File: pkg/certificate/providers/vault/types.go [hunk 1]
```diff
@@ -2,36 +2,16 @@
...
-	"time"
-
 	"github.com/hashicorp/vault/api"
-
-	"github.com/openservicemesh/osm/pkg/certificate"
-	"github.com/openservicemesh/osm/pkg/configurator"
-	"github.com/openservicemesh/osm/pkg/messaging"
 )
 
 // CertManager implements certificate.Manager and contains a Hashi Vault client instance.
```
Comment (jaellio) [inline pkg/certificate/providers/vault/types.go:8]: `CertManager` now implements the `client` interface.

File: pkg/certificate/providers/vault/certificate_manager.go [hunk 2]
```diff
@@ -30,23 +25,13 @@ const (
...
-// NewCertManager implements certificate.Manager and wraps a Hashi Vault with methods to allow easy certificate issuance.
-func NewCertManager(
-	vaultAddr,
-	token string,
-	role string,
-	cfg configurator.Configurator,
-	serviceCertValidityDuration time.Duration,
-	msgBroker *messaging.Broker) (*CertManager, error) {
+// New will construct a new certificate client using Vault's cert-manager
+func New(vaultAddr, token string, role string) (*CertManager, error) {
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager.go:32]: nit: if you're specifying the param type once for 2 variables, may as well do once for all 3

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 3]
```diff
@@ -1,88 +1,18 @@
...
-				vaultAddr,
-				vaultToken,
-				vaultRole,
-				mockConfigurator,
-				mockConfigurator.GetServiceCertValidityPeriod(),
-				nil,
-			)
-			Expect(err).To(HaveOccurred())
-			vaultError := err.(*url.Error)
-			expected := `unsupported protocol scheme "foo"`
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:82]: this test is. valid, we can keep it

**Review by steeling — APPROVED** [review]

File: pkg/certificate/providers/vault/certificate_manager.go (hunk 2 above)
Comment (schristoff) [reply pkg/certificate/providers/vault/certificate_manager.go:32]: I actually love this nit, good call out.

File: pkg/certificate/providers/vault/certificate_manager_test.go (hunk 3 above)
Comment (shashankram) [reply pkg/certificate/providers/vault/certificate_manager_test.go:82]: +1, please add this back

File: pkg/certificate/providers/vault/types.go [hunk 4]
```diff
@@ -2,36 +2,16 @@
...
 	role vaultRole
-
-	cfg configurator.Configurator
-
-	serviceCertValidityDuration time.Duration
-
-	msgBroker *messaging.Broker
 }
 
 type vaultRole string
```
Comment (steeling) [inline pkg/certificate/providers/vault/types.go:17]: the benefit of aliasing types is typically to reduce the allowed set of values. ie:

```
type vaultRole string // unexported, so needs to be specified

var (
   defaultRole vaultRole = "default-role"
)
```

Since we're not doing that, let's just go ahead and get rid of these and call them strings.

File: pkg/certificate/providers/vault/tools.go [hunk 5]
```diff
@@ -11,12 +11,8 @@ func getDurationInMinutes(validityPeriod time.Duration) string {
 	return fmt.Sprintf("%dh", validityPeriod/time.Hour)
 }
 
-func getIssueURL(role vaultRole) vaultPath {
-	return vaultPath(fmt.Sprintf("pki/issue/%+v", role))
-}
-
-func getRoleConfigURL(role vaultRole) vaultPath {
-	return vaultPath(fmt.Sprintf("pki/roles/%s", role))
+func getroleURL(role string) string {
```
Comment (steeling) [inline pkg/certificate/providers/vault/tools.go:14]: Did you intend to make change?  The functionality should stay the same (as well. as the func names), with simply changing the type from. vaultRole to string

File: pkg/certificate/providers/vault/certificate_manager.go [hunk 6]
```diff
@@ -60,42 +45,12 @@ func NewCertManager(
...
-	cm.ReleaseCertificate(cert.GetCommonName())
-
-	return issuingCA, cert.GetSerialNumber(), err
-}
-
-func (cm *CertManager) issue(cn certificate.CommonName, validityPeriod time.Duration) (*certificate.Certificate, error) {
-	secret, err := cm.client.Logical().Write(getIssueURL(cm.role).String(), getIssuanceData(cn, validityPeriod))
+// IssueCertificate will request a new signed certificate from the configured Vault issuer.
+func (cm *CertManager) IssueCertificate(cn certificate.CommonName, validityPeriod time.Duration) (*certificate.Certificate, error) {
+	secret, err := cm.client.Logical().Write(getroleURL(cm.role), getIssuanceData(cn, validityPeriod))
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager.go:53]: this is returning a new URL which would be a bug.

File: pkg/certificate/providers/vault/tools.go [hunk 7]
```diff
@@ -11,12 +11,8 @@ func getDurationInMinutes(validityPeriod time.Duration) string {
 	return fmt.Sprintf("%dh", validityPeriod/time.Hour)
 }
 
-func getIssueURL(role vaultRole) vaultPath {
-	return vaultPath(fmt.Sprintf("pki/issue/%+v", role))
-}
-
-func getRoleConfigURL(role vaultRole) vaultPath {
```
Comment (steeling) [inline pkg/certificate/providers/vault/tools.go:18]: note: this is valid to remove since it is never used. We can delete the corresponding test as well

File: pkg/certificate/providers/vault/tools_test.go [hunk 8]
```diff
@@ -25,16 +25,16 @@ var _ = Describe("Test tools", func() {
...
 		It("creates the URL for issuing a new certificate", func() {
-			actual := getIssueURL(role)
-			expected := vaultPath(fmt.Sprintf("pki/issue/%s", role))
+			actual := getroleURL(role)
+			expected := fmt.Sprintf("pki/issue/%s", role)
 			Expect(actual).To(Equal(expected))
 		})
 	})
 
 	Context("Test role config URL", func() {
```
Comment (steeling) [inline pkg/certificate/providers/vault/tools_test.go:34]: You can delete this whole test

File: pkg/certificate/providers/vault/tools.go (hunk 5 above)
Comment (schristoff) [reply pkg/certificate/providers/vault/tools.go:14]: 🤦‍♀️ went too hard on the cleanup and thought "getIssueURL" and "getRoleConfigURL" were the same thing. my b!

File: pkg/certificate/providers/vault/certificate_manager.go [hunk 9]
```diff
@@ -30,72 +25,36 @@ const (
...
-		msgBroker:                   msgBroker,
+		role: role,
 	}
 	config := api.DefaultConfig()
 	config.Address = vaultAddr
 
 	var err error
 	if c.client, err = api.NewClient(config); err != nil {
-		return nil, errors.Errorf("Error creating Vault CertManager without TLS at %s", vaultAddr)
+		return nil, errors.Errorf("Error creating Vault CertManager without TLS at %s", c.client.Address())
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager.go:45]: this could result in nil pointer dereference if an err is returned and client is nil, this should be vaultAddr

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 10]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
+		{description: "no err", wantErr: false},
+		{description: "all nil", wantErr: true},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			if tc.wantErr {
+				_, err := New("", "", "")
+				if err == nil {
+					t.Fatalf("expected error, got nil")
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:70]: Fatal prevents future tests from running, this should be an error

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 11]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
+				}
+				_, err = New(addr, token, vaultRole)
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:72]: You're actually running this inside the wantErr as well it looks like

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 12]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
+		description string
+		wantErr     bool
+	}{
+		{description: "no err", wantErr: false},
+		{description: "all nil", wantErr: true},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.description, func(t *testing.T) {
+			if tc.wantErr {
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:67]: doesn't look like. your testing the success case, so table driven tests aren't needed here.

Typically in a TDT, you would add the inputs to the test case struct.

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 13]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
-			certBytes := uuid.New().String()
-			issue := func(certificate.CommonName, time.Duration) (*certificate.Certificate, error) {
-				cert := &certificate.Certificate{
-					IssuingCA:    pem.RootCertificate(certBytes),
-					SerialNumber: certSerialNum,
+	testCases := []struct {
+		description string
+		wantErr     bool
+	}{
+		{description: "no err", wantErr: false},
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:61]: Instead of err and no err, let's provide the description of why it fails, ie:

"errors when an invalid address is provided"

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 14]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
+}
 
-		It("gets certs from cache", func() {
-			// This cert does not exist - returns nil
-			Expect(cm.getFromCache("nothing")).To(BeNil())
-
-			// This cert has expired -- returns nil
-			Expect(cm.getFromCache(expiredCertCN)).To(BeNil())
+func Test_IssueCertificate(t *testing.T) {
+	var common_name certificate.CommonName = "localhost"
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:82]: No snake case in go

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 15]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
-			Expect(actual).To(Equal(rootCert))
-		})
+func mockVault(t *testing.T) (*vault.Core, string, string) {
+	coreConfig := &vault.CoreConfig{
+		LogicalBackends: map[string]logical.Factory{
+			"pki": pki.Factory,
+		},
+	}
+	core, _, rootToken := vault.TestCoreUnsealedWithConfig(t, coreConfig)
+	_, addr := vhttp.TestServer(t, core)
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:125]: We should eventually call Close on the. returned listener. 

can we rework this to use golang's test setup?

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 16]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
+	}
+}
 
-		It("gets certs from cache", func() {
-			// This cert does not exist - returns nil
-			Expect(cm.getFromCache("nothing")).To(BeNil())
-
-			// This cert has expired -- returns nil
-			Expect(cm.getFromCache(expiredCertCN)).To(BeNil())
+func Test_IssueCertificate(t *testing.T) {
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:81]: No snake case

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 17]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
+	if err != nil {
+		t.Fatalf("did not expect error, got %v", err)
+	}
+	mockVaultwithPKI(t, cm)
+
+	testCases := []struct {
+		description string
+		wantErr     bool
+	}{
+		{description: "no err", wantErr: false},
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:98]: same notes as above

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 18]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
-			Expect(actual).To(Equal(validCert))
-		})
+}
 
-		It("can fetch root cert", func() {
-			actual, err := cm.GetRootCertificate()
-			Expect(err).ToNot(HaveOccurred())
-			Expect(actual).To(Equal(rootCert))
-		})
+func mockVault(t *testing.T) (*vault.Core, string, string) {
```
[the remark this answers is not in the document]
Comment (schristoff) [reply pkg/certificate/providers/vault/certificate_manager_test.go:118]: That's Vault running - why do you find them not useful to what we're testing?

File: pkg/certificate/providers/vault/certificate_manager.go [hunk 19]
```diff
@@ -30,72 +25,36 @@ const (
...
-	vaultAddr,
-	token string,
-	role string,
-	cfg configurator.Configurator,
-	serviceCertValidityDuration time.Duration,
-	msgBroker *messaging.Broker) (*CertManager, error) {
+// New will construct a new certificate client using Vault's cert-manager
+func New(vaultAddr, token, role string) (*CertManager, error) {
+	if role == "" {
+		return nil, errors.Errorf("Error creating Vault CertManager, expected role got %s", role)
```
[the remark this answers is not in the document]
Comment (schristoff) [reply pkg/certificate/providers/vault/certificate_manager.go:34]: good catch

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 20]
```diff
@@ -114,99 +48,115 @@ var _ = Describe("Test client helpers", func() {
...
 
-		It("implements Certificate correctly", func() {
-			actual, err := cm.GetCertificate(validCertCN)
-			Expect(err).ToNot(HaveOccurred())
-			Expect(actual.CommonName).To(Equal(validCertCN))
-			Expect([]byte(actual.GetCertificateChain())).To(Equal([]byte(validCert.GetCertificateChain())))
-			Expect(actual.GetExpiration()).To(Equal(validCert.GetExpiration()))
-			Expect([]byte(actual.IssuingCA)).To(Equal([]byte(issuingCA)))
-		})
+func mockVaultwithPKI(t *testing.T, cm *CertManager) {
```
[the remark this answers is not in the document]
Comment (schristoff) [reply pkg/certificate/providers/vault/certificate_manager_test.go:129]: This does do similar things to `mockVault()`
The insight behind my decision is for testing `New()` it doesn't matter if the PKI secret's engine for Vault is turned on or not - so I moved enabling Vault's PKI side into a separate function so it doesn't need to be used for `New()`. 

Basically `mockVault()` right now is an agent running that isn't configured at all, just hanging out.
`mockVaultwithPKI` takes what we made in `mockVault()` and configures that agent.

File: pkg/certificate/providers/vault/certificate_manager_test.go (hunk 12 above)
Comment (schristoff) [reply pkg/certificate/providers/vault/certificate_manager_test.go:67]: `doesn't look like. your testing the success case` - this was an accident 

Yeah,  I was flip flopping around on how to test this in a normal way. I could just add input and actually test what happens if we send in stuff, which _could_ be more useful than "nil" and "not nil". If you don't think so though I can take out the tables.

File: pkg/certificate/providers/vault/certificate_manager.go [hunk 21]
```diff
@@ -30,72 +25,36 @@ const (
...
 
-// NewCertManager implements certificate.Manager and wraps a Hashi Vault with methods to allow easy certificate issuance.
-func NewCertManager(
-	vaultAddr,
-	token string,
-	role string,
-	cfg configurator.Configurator,
-	serviceCertValidityDuration time.Duration,
-	msgBroker *messaging.Broker) (*CertManager, error) {
+// New will construct a new certificate client using Vault's cert-manager
```
Comment (jaellio) [inline pkg/certificate/providers/vault/certificate_manager.go:31]: nit:
```suggestion
// New constructs a new certificate client using Vault's cert-manager
```

File: pkg/certificate/providers/vault/certificate_manager.go [hunk 22]
```diff
@@ -30,72 +25,36 @@ const (
...
-
-	// We are not going to need this certificate - remove it
-	cm.ReleaseCertificate(cert.GetCommonName())
-
-	return issuingCA, cert.GetSerialNumber(), err
-}
-
-func (cm *CertManager) issue(cn certificate.CommonName, validityPeriod time.Duration) (*certificate.Certificate, error) {
-	secret, err := cm.client.Logical().Write(getIssueURL(cm.role).String(), getIssuanceData(cn, validityPeriod))
+// IssueCertificate will request a new signed certificate from the configured Vault issuer.
```
Comment (jaellio) [inline pkg/certificate/providers/vault/certificate_manager.go:55]: nit:
```suggestion
// IssueCertificate requests a new signed certificate from the configured Vault issuer.
```

File: pkg/certificate/providers/vault/certificate_manager_test.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/certificate/providers/vault/certificate_manager_test.go:129]: FWIW I agree with Jon, I think we can simplify to have the single thing running. mostly because YAGNI :)

In general modularity is good, so I see where you're coming from as well!

File: pkg/certificate/providers/vault/certificate_manager_test.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (nojnhuh) [reply pkg/certificate/providers/vault/certificate_manager_test.go:118]: Any log that's not an error seems to indicate that Vault is running as expected, so that seems like pure noise to me. I imagine if there's an easy way to turn them off it should be simple to turn them on again when debugging these tests locally if there's any doubt Vault isn't working right where those extra logs would be helpful.

Again, not a big deal though if there's no obvious way to adjust the logs.

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 23]
```diff
@@ -51,26 +52,70 @@ var _ = Describe("Test client helpers", func() {
...
+			role:        vaultRole,
+			token:       "",
+			wantErr:     true,
+		},
 	}
 
 	for _, tc := range testCases {
 		t.Run(tc.description, func(t *testing.T) {
+			_, err := New(tc.vaultaddr, tc.token, tc.role)
 			if tc.wantErr {
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:113]: nit:

```
if (err != nil) != tc.wantErr {
  t.Errorf(...)
}
```

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 24]
```diff
@@ -115,15 +173,15 @@ func TestIssueCertificate(t *testing.T) {
...
 	coreConfig := &vault.CoreConfig{
 		LogicalBackends: map[string]logical.Factory{
 			"pki": pki.Factory,
 		},
 	}
 	core, _, rootToken := vault.TestCoreUnsealedWithConfig(t, coreConfig)
-	_, addr := vhttp.TestServer(t, core)
-	return core, rootToken, addr
+	ln, addr := vhttp.TestServer(t, core)
+	return core, rootToken, addr, ln
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:184]: since you're passing in t, you can do:


```
mockVault(...) (string, string) {
   ...
   t.Cleanup(func(){
       ln.Close()
       core.Shutdown()
   })
   return rootToken, addr
}
```

**Review by steeling — CHANGES_REQUESTED** [review]

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 25]
```diff
@@ -114,99 +49,172 @@ var _ = Describe("Test client helpers", func() {
...
+			vaultaddr:   addr,
+			role:        vaultRole,
+			token:       "",
+			wantErr:     true,
+		},
+		{
+			description: "bad addr",
+			vaultaddr:   "127.1.2.3",
+			role:        vaultRole,
+			token:       "",
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:105]: This test will fail due to no token

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 26]
```diff
@@ -169,8 +169,14 @@ func mockVault(t *testing.T) (string, string) {
 	ln, addr := vhttp.TestServer(t, core)
 
 	t.Cleanup(func() {
-		ln.Close()
-		core.Shutdown()
+		err := ln.Close()
+		if err != nil {
+			t.Fatalf("could not close net listener")
```
Comment (steeling) [inline pkg/certificate/providers/vault/certificate_manager_test.go:174]: Probably t.error so we best effort shutdown core vs bail early

**Review by steeling — APPROVED** [review]

File: pkg/certificate/providers/vault/certificate_manager.go [hunk 27]
```diff
@@ -30,180 +25,65 @@ const (
...
-	cm.msgBroker.GetCertPubSub().Pub(events.PubSubMessage{
-		Kind:   announcements.CertificateRotated,
-		NewObj: newCert,
-		OldObj: oldCert.(*certificate.Certificate),
-	}, announcements.CertificateRotated.String())
-
-	log.Debug().Msgf("Rotated certificate (old SerialNumber=%s) with new SerialNumber=%s took %+v", oldCert.(*certificate.Certificate).GetSerialNumber(), newCert.SerialNumber, time.Since(start))
-
-	return newCert, nil
+	//TODO(2068): implement a delete cert
```
Comment (shashankram) [inline pkg/certificate/providers/vault/certificate_manager.go:79]: https://github.com/openservicemesh/osm/issues/2068 is closed. Does the TODO still apply?

**Review by shashankram — APPROVED** [review]

File: pkg/certificate/providers/vault/certificate_manager_test.go [hunk 28]
```diff
@@ -52,6 +53,7 @@ var _ = Describe("Test client helpers", func() {
 })
 
 func TestNew(t *testing.T) {
+	tassert := assert.New(t)
```
Comment (nojnhuh) [inline pkg/certificate/providers/vault/certificate_manager_test.go:56]: Let's move this into the `t.Run` so the error message shows exactly which test case failed. As-is the error message looks like this:
```
--- FAIL: TestNew (0.01s)
    certificate_manager_test.go:107:
                Error Trace:    certificate_manager_test.go:107
                Error:          An error is expected but got nil.
                Test:           TestNew
                Messages:       expected error, got nil
```

But with that change it would be:
```
--- FAIL: TestNew (0.01s)
    --- FAIL: TestNew/valid_inputs (0.00s)
        certificate_manager_test.go:107:
                Error Trace:    certificate_manager_test.go:107
                Error:          An error is expected but got nil.
                Test:           TestNew/valid_inputs
                Messages:       expected error, got nil
```

File: pkg/certificate/providers/config.go [hunk 29]
```diff
@@ -296,19 +296,28 @@ func (c *Config) getHashiVaultOSMCertificateManager(options VaultOptions) (certi
...
-		c.msgBroker,
 	)
 	if err != nil {
 		return nil, nil, errors.Errorf("Error instantiating Hashicorp Vault as a Certificate Manager: %+v", err)
 	}
 
-	return vaultCertManager, vaultCertManager, nil
+	vaultCert, err := vaultClient.GetRootCertificate()
+	if err != nil {
+		return nil, nil, errors.Errorf("Error getting Vault Root Certificate, got: %v", err)
```
[the remark this answers is not in the document]
Comment (steeling) [reply pkg/certificate/providers/config.go]: https://github.com/openservicemesh/osm/issues/4524 :)
